lib/matplotlib/cbook/__init__.py

```diff
@@ -2194,6 +2194,48 @@ def _check_in_list(_values, **kwargs):
                 .format(v, k, ', '.join(map(repr, values))))
 
 
+def _check_shape(_shape, **kwargs):
+    """
+    For each *key, value* pair in *kwargs*, check that *value* has the shape
+    *_shape*, if not, raise an appropriate ValueError.
+
+    None in the shape is treated as a "free" sizes that can have any length.
+    e.g. (None, 2) -> (N, 2)
+
+    The values checked must be numpy arrays.
+
+    Examples
+    --------
+    To check from (N, 2) shaped arrays
+
+    >>> cbook._check_in_list((None, 2), arg=arg, other_arg=other_arg)
+    """
+    target_shape = _shape
+    for k, v in kwargs.items():
+        data_shape = v.shape
+
+        if not (
+            len(data_shape) == len(target_shape)
+            and all(
+                (t == s if t is not None else True)
+                for t, s in zip(target_shape, data_shape)
+            )
+        ):
+            def format_dims(target_shape):
+                dim_labels = iter(itertools.chain(
+                    'MNLIJKLH',
+                    (f"D{i}" for i in itertools.count())))
+                text_shape = tuple(n if n is not None else next(dim_labels)
+                                   for n in target_shape)
+                return '(' + ", ".join(str(_) for _ in text_shape) + ')'
```

Comment:
fstrings are great but I don't think this warrants being its own inner function?

Reply:
fair